### ISSUE TITLE
drms: sizeratio keyword for im_patch cutouts is not implemented/working

### ISSUE THREAD

**Issue by Tom-Berger** [sunpy/drms#57]
### DRMS ignoring sizeratio keyword?

When attempting to download large numbers of AIA images using the im_patch process, drms apparently does not implement the "sizeratio" keyword that enables JSOC to recognize the true total data size requested. The request is apparently immediately rejected if the size of the associated _full-disk_ request exceeds the 100GB limit. Here's an example:
```python
qstr = 'aia.lev1_euv_12s[2017.08.31_09:00:00-2017.09.10_11:00:00@48s][131][?quality>=0?]{image}'
process = {'im_patch': { 't_ref': trefi, 't': 0, 'r': 1, 'c': 0, 
                                                 'locunits': 'stony',
                                                 'cadence':48,            
                                                 'boxunits': 'pixels', 'x': lon_center, 'y': lat_center,
                                                 'width': nx,'height': ny,  'sizeratio':sizeratio                      
                        }
            }

result = client.export(qstr, method='url', protocol='fits', email=email, process=process)
result.request_url(verbose=FALSE)
```

This is a request for 18,105 AIA 131 images cut out to match the HMI SHARPs 7115 cutout. The size of the (compressed) full disk fits file set is about 144 GB so the result is the following error:
```
DrmsExportError: Request exceeds max byte limit of 100000MB [status=3]
```

However, when submitting this same request via JSOC export_data via the following command:

```
op=exp_request&ds=aia.lev1_euv_12s[2017.08.31_09:00:00-2017.09.10_11:00:00@48s][131][?quality>=0?]&sizeratio=0.01837158203125&process=n=0|im_patch,t_start=2017-08-31T08:59:59Z,t_stop=2017-09-10T10:59:59Z,t=0,r=0,c=0,cadence=48s,locunits=stony,boxunits=pixels,t_ref=2017-08-31T08:59:59Z,x=-44.71,y=-8.66,width=688,height=448&requestor=JSOC_Export&notify=Thomas.berger@colorado.edu&method=url&filenamefmt=aia.lev1_euv_12s.{T_REC:A}.{WAVELNTH}.{segment}&format=json&protocol=FITS,compress Rice
```

The result is a request for 3763 MB of data which is fulfilled. JSOC uses the sizeratio keyword to compute the actual data volume prior to queuing up the request for download. Note that sizeratio here is 0.018 since the cut out is only 688x448 pixels. 

This lack of implementation of the sizeratio keyword (or bug in the implementation?) is a major hinderance to downloading AIA data for machine learning purposes where we naturally need to download 100s of thousands of images for sufficient training diversity.

**Comment by davisite** [comment]
Here's more info:

To address this issue, I created code that is OPTIONALLY invoked when the `info` JSOC endpoint (the jsoc_info URL) is opened. When invoked, this code returns the size ratio that results when applying a pipeline of processing programs to an input drms series. I modified Sunpy/drms to open this endpoint when the Sunpy/drms user wishes to have DRMS calculate the size ratio. This size ratio is then passed to the existing `fetch` JSOC endpoint (the jsoc_fetch URL). If a user calls the Client.export() API method with a `process` dict, and a new size-ratio flag, that dict is used to create an ordered list of processing programs that, along with the series, is sent to the `info` endpoint. But to make this work, I had to address two other issues.

First, the export processing steps that Sunpy/drms enumerates are not quite the same as the processing-program pipeline needed by the `info` endpoint. When a user wishes to specify a particular set of export-processing steps, they specify the export programs, the arguments of each export program, PLUS the pipeline arguments for each program. Accordingly, there are eight export-processing "types": 'aia_scale_orig', 'aia_scale_aialev1', 'aia_scale_other', 'HmiB2ptr', 'resize', 'im_patch', 'Maproj', and 'rebin', each of which identifies pipeline arguments (there is a mapping between these names and the pipeline arguments that exists at the JSOC). But when a user wishes to determine the size ratio, they need only specify an ordered list of a subset of the six export programs, plus their arguments: 'generate_aia_level15', 'convert_field_components', 'resize_image', 'extract_patches', 'map_to_projection', and 'rebin_image' (these are new unique names I chose - their names do not clash with any of the processing-step type names or anything else). It is possible to map the eight 'export' processing types to the six processing programs, and it is possible to perform the reverse mapping, provided that the series is known.

Second, the order in which export processing steps are applied matters. The final images resulting from applying steps in one order might differ from those resulting from the application of the steps in a different order. The export system at the JSOC enforces a specific order. Unfortunately, the order is specified in the export web app, which means that if we want Sunpy/drms to use the same order, then Sunpy/drms has to also specify the order. But currently, Sunpy/drms iterates through the processing dict and the order that the steps are iterated through becomes that order. Since a dict is an unordered collection of data, Sunpy/drms is not currently ordering the processing steps, not in a way that matches what a user would expect had they used the JSOC export system. I consider this lack of order a Sunpy/drms bug.

To deal with both issues I created a new class `ProcessingSteps`. It solves the first issue by managing the mapping of the eight export-processing types to the six size-processing steps (and the reverse too). In this way, a user can continue to provide the same `process` dict to the Client.export() call, and if they also supply a new flag to invoke the size-ratio code, then this mapping will be used to then create the list of programs that are sent to the `info` endpoint. It solves the second issue by having a class variable that has a list (an ordered collection) of processing programs. Right before the `fetch` URL is created, the `process` dict is ordered using the list in this class variable.

**Comment by davisite** [comment]
Tom and I are testing this all out; it turns out there have been other JSOC export-system bugs not related to the new code (which are hopefully fixed now) interfering.

**Comment by davisite** [comment]
Now that I've wrote a detailed description, I've been thinking about this, and I think I'm going to explore a better way of specifying the sequence of processing steps - why not do this right and have the server define a default sequence (the one currently in use by the JSOC export web app), and give the clients (export web app, Sunpy/drms) the ability to override this.

**Comment by wtbarnes** [comment]
Hi @davisite or @Tom-Berger, sorry for not following up on this sooner. Has any progress been made on this front? Is there a potential fix for this that we could start iterating on?

**Comment by davisite** [comment]
Hi Will,

It has been a while since I’ve looked at this; I believe Tom has been making progress.

My fix for this was to put the code that calculates the size ratio on the server side – having it on the client side and then passing it to the server never really made sense since the server has all the info needed to do the calculation.

I believe I put this fix into an edited drms, but I never did anything beyond that.

Art

**Comment by Tom-Berger** [comment]
I can confirm that Art’s fix has been working for us. Case closed as far as I’m concerned.

Thanks again for everyone’s help on this.
Tom


From: Arthur Amezcua ***@***.***>
Date: Thursday, September 23, 2021 at 9:41 AM
To: sunpy/drms ***@***.***>
Cc: Tom Berger ***@***.***>, Mention ***@***.***>
Subject: Re: [sunpy/drms] drms: sizeratio keyword for im_patch cutouts is not implemented/working (#57)
Hi Will,

It has been a while since I’ve looked at this; I believe Tom has been making progress.

My fix for this was to put the code that calculates the size ratio on the server side – having it on the client side and then passing it to the server never really made sense since the server has all the info needed to do the calculation.

I believe I put this fix into an edited drms, but I never did anything beyond that.

Art

**Comment by wtbarnes** [comment]
As it looks like Art's upstream fix has solved Tom's original issue, I'm going to close this.
